src/services/mapCode.ts

```diff
@@ -0,0 +1,341 @@
+        );
+        if (scope) {
+            const start = scope.statements.find(stmt => changes.some(node => matchNode(node, stmt)));
+            if (start) {
+                // Can't be undefined here, since we know we have at least one match.
+                const end = findLast(scope.statements, stmt => changes.some(node => matchNode(node, stmt)))!;
+                forEach(changes, wipeNode);
+                changeTracker.replaceNodeRangeWithNodes(
+                    originalFile,
+                    start,
+                    end,
+                    changes,
+                );
+                return;
+            }
+        }
+    }
+
+    let scopeStatements: NodeArray<Statement> = originalFile.statements;
+    for (const location of focusLocations) {
+        const block = findAncestor(
+            getTokenAtPosition(originalFile, location.start),
+            isBlock,
+        );
+        if (block) {
+            scopeStatements = block.statements;
+            break;
+        }
+    }
+    forEach(changes, wipeNode);
+    changeTracker.insertNodesAfter(
+        originalFile,
+        scopeStatements[scopeStatements.length - 1],
+        changes,
+    );
+}
+
+function matchNode(a: Node, b: Node): boolean {
+    if (a.kind !== b.kind) {
+        return false;
+    }
+
+    if (a.kind === SyntaxKind.Constructor) {
+        return a.kind === b.kind;
+    }
+
+    if (isNamedDeclaration(a) && isNamedDeclaration(b)) {
+        return a.name.getText() === b.name.getText();
+    }
+
+    if (isIfStatement(a) && isIfStatement(b)) {
+        return (
+            a.expression.getText() === b.expression.getText()
+        );
+    }
+
+    if (isWhileStatement(a) && isWhileStatement(b)) {
+        return (
+            a.expression.getText() ===
+                b.expression.getText()
+        );
+    }
+
+    if (isForStatement(a) && isForStatement(b)) {
+        return (
+            a.initializer?.getText() ===
+                b.initializer?.getText() &&
+            a.incrementor?.getText() ===
+                b.incrementor?.getText() &&
+            a.condition?.getText() === b.condition?.getText()
+        );
+    }
+
+    if (isForInOrOfStatement(a) && isForInOrOfStatement(b)) {
+        return (
+            a.expression.getText() ===
+                b.expression.getText() &&
+            a.initializer.getText() ===
+                b.initializer.getText()
+        );
+    }
+
+    if (isLabeledStatement(a) && isLabeledStatement(b)) {
+        // If we're actually labeling/naming something, we should be a bit
+        // more lenient about when we match, so we don't care what the actual
+        // related statement is: we just replace.
+        return a.label.getText() === b.label.getText();
+    }
+
+    if (a.getText() === b.getText()) {
+        return true;
+    }
+
+    return false;
+}
+
+function wipeNode(node: Mutable<Node>) {
+    resetNodePositions(node);
+    node.parent = undefined!;

```

Comment:
this is from a SourceFile, but the SourceFile is "fake". I need to wipe it/replace its parent so that node manipulation/insertion stuff doesn't get confused, so this is necessary.